### ISSUE TITLE
Flaky Rolling Update test

### ISSUE THREAD

**Issue by rifelpet** [kubernetes/kops#9040]
I noticed in an open PR of mine that a rolling update test failed in a GitHub Actions job. I'm not sure if its a timing issue due to the job's limited resources or what.

I don't see a way to view previous builds of a GHA job and the URL to the raw job output uses a short term token that has expired already, so here is the output of the test:

```
2020-05-02T12:20:03.0900200Z === RUN   TestRollingUpdateMaxSurgeAllNeedUpdateMaxAlreadyDetached
2020-05-02T12:20:03.0955890Z I0502 12:19:55.078416    3865 instancegroups.go:290] Tainting 7 nodes in "node-1" instancegroup.
2020-05-02T12:20:03.0992110Z I0502 12:19:55.079323    3865 instancegroups.go:348] Draining the node: "node-1a.local".
2020-05-02T12:20:03.1045900Z I0502 12:19:55.079511    3865 instancegroups.go:367] deleting node "node-1a.local" from kubernetes
2020-05-02T12:20:03.1304340Z I0502 12:19:55.079528    3865 instancegroups.go:517] Stopping instance "node-1a", node "node-1a.local", in group "node-1" (this may take a while).
2020-05-02T12:20:03.1456860Z I0502 12:19:55.079542    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.1557830Z I0502 12:19:55.088101    3865 instancegroups.go:391] Validating the cluster.
2020-05-02T12:20:03.1738020Z I0502 12:19:55.088145    3865 instancegroups.go:432] Cluster validated.
2020-05-02T12:20:03.1819370Z I0502 12:19:55.088165    3865 instancegroups.go:348] Draining the node: "node-1e.local".
2020-05-02T12:20:03.1968380Z I0502 12:19:55.088453    3865 instancegroups.go:367] deleting node "node-1e.local" from kubernetes
2020-05-02T12:20:03.2053920Z I0502 12:19:55.088472    3865 instancegroups.go:517] Stopping instance "node-1e", node "node-1e.local", in group "node-1" (this may take a while).
2020-05-02T12:20:03.2168390Z I0502 12:19:55.088496    3865 instancegroups.go:348] Draining the node: "node-1c.local".
2020-05-02T12:20:03.2387210Z I0502 12:19:55.088672    3865 instancegroups.go:367] deleting node "node-1c.local" from kubernetes
2020-05-02T12:20:03.2494470Z I0502 12:19:55.088688    3865 instancegroups.go:517] Stopping instance "node-1c", node "node-1c.local", in group "node-1" (this may take a while).
2020-05-02T12:20:03.2655990Z I0502 12:19:55.088701    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.2757220Z I0502 12:19:55.108794    3865 instancegroups.go:391] Validating the cluster.
2020-05-02T12:20:03.2820830Z I0502 12:19:55.108846    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.2878170Z I0502 12:19:55.132805    3865 instancegroups.go:432] Cluster validated.
2020-05-02T12:20:03.2922750Z I0502 12:19:55.132845    3865 instancegroups.go:348] Draining the node: "node-1f.local".
2020-05-02T12:20:03.3024630Z I0502 12:19:55.133153    3865 instancegroups.go:367] deleting node "node-1f.local" from kubernetes
2020-05-02T12:20:03.3127130Z I0502 12:19:55.133167    3865 instancegroups.go:517] Stopping instance "node-1f", node "node-1f.local", in group "node-1" (this may take a while).
2020-05-02T12:20:03.3204660Z I0502 12:19:55.133188    3865 instancegroups.go:391] Validating the cluster.
2020-05-02T12:20:03.3360550Z I0502 12:19:55.133219    3865 instancegroups.go:432] Cluster validated.
2020-05-02T12:20:03.3364740Z I0502 12:19:55.133230    3865 instancegroups.go:348] Draining the node: "node-1g.local".
2020-05-02T12:20:03.3439300Z I0502 12:19:55.133404    3865 instancegroups.go:367] deleting node "node-1g.local" from kubernetes
2020-05-02T12:20:03.3533370Z I0502 12:19:55.133415    3865 instancegroups.go:517] Stopping instance "node-1g", node "node-1g.local", in group "node-1" (this may take a while).
2020-05-02T12:20:03.3642720Z I0502 12:19:55.133502    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.3744020Z I0502 12:19:55.134693    3865 instancegroups.go:391] Validating the cluster.
2020-05-02T12:20:03.3845150Z I0502 12:19:55.134767    3865 instancegroups.go:432] Cluster validated.
2020-05-02T12:20:03.3923810Z I0502 12:19:55.134781    3865 instancegroups.go:348] Draining the node: "node-1b.local".
2020-05-02T12:20:03.3993400Z I0502 12:19:55.134991    3865 instancegroups.go:367] deleting node "node-1b.local" from kubernetes
2020-05-02T12:20:03.4106950Z I0502 12:19:55.135003    3865 instancegroups.go:517] Stopping instance "node-1b", node "node-1b.local", in group "node-1" (this may take a while).
2020-05-02T12:20:03.4122950Z I0502 12:19:55.135014    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.4124250Z I0502 12:19:55.136195    3865 instancegroups.go:391] Validating the cluster.
2020-05-02T12:20:03.4124800Z I0502 12:19:55.136215    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.4125310Z I0502 12:19:55.179822    3865 instancegroups.go:432] Cluster validated.
2020-05-02T12:20:03.4126500Z I0502 12:19:55.180195    3865 instancegroups.go:348] Draining the node: "node-1d.local".
2020-05-02T12:20:03.4127510Z I0502 12:19:55.180800    3865 instancegroups.go:367] deleting node "node-1d.local" from kubernetes
2020-05-02T12:20:03.4128540Z I0502 12:19:55.180832    3865 instancegroups.go:517] Stopping instance "node-1d", node "node-1d.local", in group "node-1" (this may take a while).
2020-05-02T12:20:03.4129110Z I0502 12:19:55.180852    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.4129620Z I0502 12:19:55.182073    3865 instancegroups.go:391] Validating the cluster.
2020-05-02T12:20:03.4130120Z I0502 12:19:55.182119    3865 instancegroups.go:432] Cluster validated.
2020-05-02T12:20:03.4130650Z I0502 12:19:55.182128    3865 rollingupdate.go:201] Rolling update completed for cluster ""!
2020-05-02T12:20:03.4131620Z --- FAIL: TestRollingUpdateMaxSurgeAllNeedUpdateMaxAlreadyDetached (0.10s)
2020-05-02T12:20:03.4133890Z ##[error]    rollingupdate_test.go:909: unexpected call to Validate with 3 termination requests left
2020-05-02T12:20:03.4150450Z     rollingupdate_test.go:976: 
2020-05-02T12:20:03.4156730Z         	Error Trace:	rollingupdate_test.go:976
2020-05-02T12:20:03.4157340Z         	            				aws_cloud.go:444
2020-05-02T12:20:03.4157880Z         	            				mock_aws_cloud.go:90
2020-05-02T12:20:03.4158410Z         	            				instancegroups.go:522
2020-05-02T12:20:03.4158960Z         	            				instancegroups.go:374
2020-05-02T12:20:03.4159500Z         	            				instancegroups.go:172
2020-05-02T12:20:03.4160020Z         	            				asm_amd64.s:1357
2020-05-02T12:20:03.4160530Z         	Error:      	Not equal: 
2020-05-02T12:20:03.4161040Z         	            	expected: 4
2020-05-02T12:20:03.4161540Z         	            	actual  : 3
2020-05-02T12:20:03.4162230Z         	Test:       	TestRollingUpdateMaxSurgeAllNeedUpdateMaxAlreadyDetached
2020-05-02T12:20:03.4163420Z         	Messages:   	previous validation
2020-05-02T12:20:03.4164170Z     rollingupdate_test.go:925: 
2020-05-02T12:20:03.4164730Z         	Error Trace:	rollingupdate_test.go:925
2020-05-02T12:20:03.4165470Z         	            				instancegroups.go:428
2020-05-02T12:20:03.4165910Z         	            				instancegroups.go:411
2020-05-02T12:20:03.4272640Z         	            				instancegroups.go:393
2020-05-02T12:20:03.4340610Z         	            				instancegroups.go:188
2020-05-02T12:20:03.4369100Z         	            				rollingupdate.go:187
2020-05-02T12:20:03.4395680Z         	            				rollingupdate_test.go:1355
2020-05-02T12:20:03.4417230Z         	Error:      	Not equal: 
2020-05-02T12:20:03.4473350Z         	            	expected: 4
2020-05-02T12:20:03.4475960Z         	            	actual  : 3
2020-05-02T12:20:03.4586250Z         	Test:       	TestRollingUpdateMaxSurgeAllNeedUpdateMaxAlreadyDetached
2020-05-02T12:20:03.4602020Z         	Messages:   	previous validation
```

whenever you get a chance,
/assign @johngmyers

**Comment by johngmyers** [comment]
Thanks, I'll take a look.

**Comment by johngmyers** [comment]
TLDR in the next comment.

Perhaps a very long GC pause? Things went off track around the lines:

```
2020-05-02T12:20:03.2820830Z I0502 12:19:55.108846    3865 instancegroups.go:380] waiting for 1ms after terminating instance
2020-05-02T12:20:03.2878170Z I0502 12:19:55.132805    3865 instancegroups.go:432] Cluster validated.
```
The test expects the following sequence of events:

main thread: start goroutines A and B, calling `drainTerminateAndWait` to terminate node-1e and node-1c.
https://github.com/kubernetes/kops/blob/df7e0b18b62599aa610e1c19bde411fd76e9fdee/pkg/instancegroups/instancegroups.go#L173-L171
Then block on `terminateChan`:
https://github.com/kubernetes/kops/blob/df7e0b18b62599aa610e1c19bde411fd76e9fdee/pkg/instancegroups/instancegroups.go#L182

The first goroutiine terminates node-1e, blocking on `terminationChan`
https://github.com/kubernetes/kops/blob/df7e0b18b62599aa610e1c19bde411fd76e9fdee/pkg/instancegroups/rollingupdate_test.go#L969

The second goroutine terminates node 1c, waits the 1ms `sleepAfterTerminate`, and signals `terminateChan`, waking up the main thread.

The main thread then calls `maybeValidate`, signaling `terminationChan` to wake up the first goroutine and then blocking on `validationChan`
https://github.com/kubernetes/kops/blob/df7e0b18b62599aa610e1c19bde411fd76e9fdee/pkg/instancegroups/rollingupdate_test.go#L919

The first goroutine starts a third goroutine and the race is on
https://github.com/kubernetes/kops/blob/df7e0b18b62599aa610e1c19bde411fd76e9fdee/pkg/instancegroups/rollingupdate_test.go#L974

The first goroutine returns to `drainTerminateAndWait` to sleep 1ms, return, and signal `terminateChan`.

Concurrently, the third goroutine sleeps 20ms then signals `validationChan` to wake the main thread

The test expects the signal from the first goroutine to be available by the time the main thread selects for it at
https://github.com/kubernetes/kops/blob/df7e0b18b62599aa610e1c19bde411fd76e9fdee/pkg/instancegroups/instancegroups.go#L214

In the flaky run, the first goroutine lost the race and that `select` took the `default` branch. That the code under test then started only one new drain before waiting was caught in subsequent assertions.

**Comment by johngmyers** [comment]
This test is complicated but it found a real bug where the code wasn’t using a labeled break and thus was blocking for all pending terminations to complete before starting any new ones. On the other hand, the sweep block is only useful for concurrency values other than 1 or 100%, which are probably going to be rare.

We can increase the delay in `delayThenWakeValidation`, but that is called 14 times throughout the test suite. There is also a corresponding delay in `alreadyDetachedTest.Validate` that is called once.  So increasing it from 20ms to 100ms would make the tests take 1.2s longer.

So what kind of time budget should these tests have to avoid flakes?
